src/main/java/uk/nhs/tis/trainee/notifications/service/ProgrammeMembershipService.java

```diff
@@ -246,67 +243,101 @@ public void addNotifications(ProgrammeMembership programmeMembership)
+      jobDataMap.put(COJ_SYNCED_FIELD, programmeMembership.getConditionsOfJoining().syncedAt());
+    }
+    jobDataMap.put(RO_NAME_FIELD, getRoName(programmeMembership.getResponsibleOfficer()));
+    jobDataMap.put(DESIGNATED_BODY_FIELD, programmeMembership.getDesignatedBody());
+  }
 
-      JobDataMap pmDayOneJobDataMap = new JobDataMap();
-      pmDayOneJobDataMap.put(TIS_ID_FIELD, programmeMembership.getTisId());
-      pmDayOneJobDataMap.put(PERSON_ID_FIELD, programmeMembership.getPersonId());
-      pmDayOneJobDataMap.put(PROGRAMME_NAME_FIELD, programmeMembership.getProgrammeName());
-      pmDayOneJobDataMap.put(PROGRAMME_NUMBER_FIELD, programmeMembership.getProgrammeNumber());
-      pmDayOneJobDataMap.put(START_DATE_FIELD, startDate);
-      pmDayOneJobDataMap.put(TEMPLATE_OWNER_FIELD, programmeMembership.getManagingDeanery());
-      pmDayOneJobDataMap.put(TEMPLATE_NOTIFICATION_TYPE_FIELD, PROGRAMME_DAY_ONE);
-      pmDayOneJobDataMap.put(RO_NAME_FIELD, getRoName(programmeMembership.getResponsibleOfficer()));
-      pmDayOneJobDataMap.put(DESIGNATED_BODY_FIELD, programmeMembership.getDesignatedBody());
-
-      String jobId = PROGRAMME_DAY_ONE + "-" + programmeMembership.getTisId();
-      if (startDate.isBefore(LocalDate.now(timezone).plusDays(1))) {
-        notificationService.executeNow(jobId, pmDayOneJobDataMap);
-      } else {
-        notificationService.scheduleNotification(jobId, pmDayOneJobDataMap,
-            Date.from(startDate.atStartOfDay(timezone).toInstant()), ONE_DAY_IN_SECONDS);
+  /**
+   * Determine when a programme notification should be scheduled.
+   *
+   * @param notificationType         The type of notification to schedule.
+   * @param programmeMembership      The programme membership to consider.
+   * @param notificationsAlreadySent The notifications already sent for this entity.
+   * @return Whether the notification is unnecessary, and the date when the notification should be
+   *     scheduled, or null if it should be sent immediately.
+   */
+  private NotificationScheduleWhen whenScheduleProgrammeNotification(
+      NotificationType notificationType, ProgrammeMembership programmeMembership,
+      Map<NotificationType, History> notificationsAlreadySent) {
+    if (notificationType == PROGRAMME_CREATED) {
+      return new NotificationScheduleWhen(false,
+          whenScheduleDeferredNotification(PROGRAMME_CREATED, programmeMembership,
+              notificationsAlreadySent));
+    } else {
+      Integer daysBeforeStart = getDaysBeforeStartForNotification(notificationType);
+      if (NotificationType.getReminderProgrammeUpdateNotificationTypes().contains(notificationType)) {
+        if (programmeMembership.getStartDate().minusDays(daysBeforeStart)
+            .isBefore(LocalDate.now(timezone))) {
+          // If the notification is a reminder and the deadline is in the past, it is unnecessary.
+          return new NotificationScheduleWhen(true, null);
```

Comment:
This is the only notification type for which overdue notifications are skipped instead of being sent immediately, hence the fiddly logic.